src/qcrypto/simbasics.py

```diff
@@ -1,132 +1,424 @@
+    def measure(self):
+        pass
+
+    @abstractmethod
+    def measure_all(self):
+        pass
+
+    @abstractmethod
+    def apply_gate(self, gate):
+        pass
+
+    @abstractmethod
+    def _calculate_measurement_probs(self, qubit_idx):
+        pass
+
+    @abstractmethod
+    def _update_state_post_measurement(self, qubit_idx, outcome):
+        pass
+
+    @abstractmethod
+    def _normalize_state(self):
+        pass
 
 
 @dataclasses.dataclass
-class Qubit:
-    """
-    Data class which represents the qubit and its quantum state.
-    Quantum state is given by:
-    |\psi> = \cos(\frac12 \theta) |0> + e^{i\phi}  \sin(\frac12 \theta) |1>
-    """
+class QstateUnEnt(QState):
+    _state: np.ndarray = None
+    num_qubits: int = 10
+    init_method: str = "zeros"
 
-    def __init__(self, theta, base, phi=0):
-        self.base = base
-        self.theta = theta  # Meaningless once projected. FIX
-        self.phi = phi  # Same with this. FIX
-        self.state = np.array([np.cos(theta / 2), np.exp(1j * phi) * np.sin(theta / 2)])
+    def __post_init__(self):
+        if self._state is None:
+            if self.init_method == "zeros":
+                self._state = np.zeros((self.num_qubits, 2), dtype=np.complex_)
+                self._state[:, 0] = 1
+            elif self.init_method == "random":
+                self._state = np.random.random(
+                    (self.num_qubits, 2)
+                ) + 1j * np.random.random((self.num_qubits, 2))
+                self._normalize_state()
+            else:
+                raise ValueError("Invalid initialization method.")
+        else:
+            if self._state.shape != (self.num_qubits, 2):
+                raise ValueError(
+                    "State vector shape not appropriate for the number of qubits."
+                )
+            self._normalize_state()
+            self.num_qubits = len(self._state)
+        self._state = np.asarray(self._state, dtype=np.complex_)
 
-    def __str__(self):
-        return "{} |0> + {} |1>".format(self.state[0], self.state[1])
+    def measure(self, qubit_idx: int = None) -> int:
+        """
+        Simulates the measurement of a single qubit. As a result, the state of said qubit is collapsed depending on the result.
 
-    def __repr__(self):
-        return "{} |0> + {} |1>".format(self.state[0], self.state[1])
+        Args:
+            qubit_idx (int): Index of the qubit to be measured
+
+        Returns:
+            Outcome of the measurement. Also collapses the state of the qubit.
+
+        """
+        if qubit_idx is not None:
+            probs_0, probs_1 = self._calculate_measurement_probs(qubit_idx)
+            outcome = np.random.choice([0, 1], p=[probs_0, probs_1])
+            self._update_state_post_measurement(qubit_idx, outcome)
+            return outcome
+        else:
+            raise ValueError("qubit_idx not specified")
+
+    def measure_all(self, *kwargs) -> npt.NDArray[int]:
```

Comment:
My bad, it should be `*args`. The purpose of this is to avoid having additional `if` statements in other parts of the code which check if the `Qstate` object is `QstateUnEnt` or `QstateEnt`, since both of these classes implement a `measure_all` method. Surely there's a better way of doing this, but this was just the fastest and most innocuous at the moment, though maybe a bit confusing on first pass.